src/tools/launcher/util/launcher_util_test.cc

```diff
@@ -87,12 +93,179 @@ TEST_F(LaunchUtilTest, BashEscapeArgTest) {
+    if (!ok) {
+      DWORD err = GetLastError();
+      ASSERT_EQ(err, 0);
+    }
+    CloseHandle(processInfo.hThread);
+    bazel::windows::AutoHandle process(processInfo.hProcess);
+
+    // Wait for the subprocess to exit. Timeout is 5 seconds, which should be
+    // more than enough for the subprocess to finish.
+    ASSERT_EQ(WaitForSingleObject(process, 5000), WAIT_OBJECT_0);
+
+    // The subprocess printed its argv[1] (without a newline) to its stdout,
+    // which is redirected into the pipe.
+    // Let's write a null-terminator to the pipe to separate the output from the
+    // output of the subsequent subprocess. The null-terminator also yields
+    // null-terminated strings in the pipe, making it easy to read them out
+    // later.
+    DWORD dummy;
+    ASSERT_TRUE(WriteFile(pipe_write, "\0", 1, &dummy, NULL));
+  }
+
+  // Read the output of the subprocesses from the pipe. They are divided by
+  // null-terminators, so 'buf' will contain a sequence of null-terminated
+  // strings.  We close the writing end so that ReadFile won't block until the
+  // desired amount of bytes is available.
+  DWORD total_output_len;
+  char buf[0x10000];
+  pipe_write = INVALID_HANDLE_VALUE;
+  if (!ReadFile(pipe_read, buf, 0x10000, &total_output_len, NULL)) {
+    DWORD err = GetLastError();
+    ASSERT_EQ(err, 0);
+  }
+
+  // Assert that the subprocesses produced exactly the *unescaped* arguments.
+  size_t start = 0;
+  for (const auto& arg : args) {
+    // Assert that there was enough data produced by the subprocesses.
+    ASSERT_LT(start, total_output_len);
+
+    // Find the output of the corresponding subprocess. Since all subprocesses
+    // printed into the same pipe and we added null-terminators between them,
+    // the output is already there, conveniently as a null-terminated string.
+    string actual_arg(buf + start);
+    start += actual_arg.size() + 1;
+
+    // 'args' contains wchar_t strings, but the subprocesses printed ASCII
+    // (char) strings. To compare, we convert arg.first to a char-string.
+    string expected_arg;
+    expected_arg.reserve(arg.first.size());
+    for (const auto& wc : arg.first) {
+      expected_arg.append(1, static_cast<char>(wc));
+    }
+
+    // Assert that the subprocess printed exactly the *unescaped* argument.
+    EXPECT_EQ(expected_arg, actual_arg);
+  }
+}
+
 TEST_F(LaunchUtilTest, WindowsEscapeArgTest) {
-  ASSERT_EQ(L"foo\\bar", WindowsEscapeArg(L"foo\\bar"));
-  ASSERT_EQ(L"C:\\foo\\bar\\", WindowsEscapeArg(L"C:\\foo\\bar\\"));
-  ASSERT_EQ(L"\"C:\\foo foo\\bar\\\"", WindowsEscapeArg(L"C:\\foo foo\\bar\\"));
+  // List of arguments with their expected WindowsEscapeArg-encoded version.
+  AssertSubprocessReceivesArgsAsIntended({
+    // Each pair is:
+    // - first: argument to pass (and expected output from subprocess) 
+    // - second: expected WindowsEscapeArg-encoded string
+    {L"foo", L"foo"},
+    {L"", L"\"\""},
+    {L"foo\\bar", L"foo\\bar"},
```

Comment:
How about a single space or a path with space? Maybe adding
 ```
{L" ", L"\" \""},
{L"C:\\foo bar\\", L"C:\\foo bar\\"}, // Or should it be L"\"C:\\foo bar\\\""? I'm not sure.
```

Reply:
Done for the first one.
The second example is handled wrong now. Added that as a comment below the TODO.